File: backend/src/models/Transaction.js
const { supabase } = require('../db/supabase');
const logger = require('../utils/logger');
const { v4: uuidv4 } = require('uuid');

class Transaction {
  constructor(data) {
    this.id = data.id;
    this.txnId = data.txn_id;
    this.userId = data.user_id;
    this.amount = data.amount;
    this.currency = data.currency || 'MYR';
    this.convertedAmount = data.converted_amount;
    this.convertedCurrency = data.converted_currency;
    this.status = data.status || 'pending';
    this.type = data.type;
    this.paymentMethod = data.payment_method;
    this.merchantId = data.merchant_id;
    this.merchantName = data.merchant_name;
    this.description = data.description;
    this.metadata = data.metadata || {};
    this.createdAt = data.created_at;
    this.updatedAt = data.updated_at;
  }

  /**
   * Create a new transaction
   */
  static async create(transactionData) {
    try {
      const txnId = transactionData.txnId || `TXN-${Date.now()}-${uuidv4().substr(0, 8)}`;
      
      const { data, error } = await supabase
        .from('transactions')
        .insert([{
          txn_id: txnId,
          user_id: transactionData.userId,
          amount: transactionData.amount,
          currency: transactionData.currency || 'MYR',
          converted_amount: transactionData.convertedAmount,
          converted_currency: transactionData.convertedCurrency,
          status: transactionData.status || 'pending',
          type: transactionData.type,
          payment_method: transactionData.paymentMethod,
          merchant_id: transactionData.merchantId,
          merchant_name: transactionData.merchantName,
          description: transactionData.description,
          metadata: transactionData.metadata || {}
        }])
        .select()
        .single();

      if (error) throw error;

      logger.logTransaction(txnId, 'created', {
        amount: transactionData.amount,
        currency: transactionData.currency,
        type: transactionData.type
      });

      return new Transaction(data);
    } catch (error) {
      logger.logError(error, { action: 'createTransaction', ...transactionData });
      throw error;
    }
  }

  /**
   * Find transaction by ID
   */
  static async findById(id) {
    try {
      const { data, error } = await supabase
        .from('transactions')
        .select('*')
        .eq('id', id)
        .single();

      if (error) {
        if (error.code === 'PGRST116') return null;
        throw error;
      }

      return data ? new Transaction(data) : null;
    } catch (error) {
      logger.logError(error, { action: 'findTransactionById', transactionId: id });
      throw error;
    }
  }

  /**
   * Find transaction by transaction ID
   */
  static async findByTxnId(txnId) {
    try {
      const { data, error } = await supabase
        .from('transactions')
        .select('*')
        .eq('txn_id', txnId)
        .single();

      if (error) {
        if (error.code === 'PGRST116') return null;
        throw error;
      }

      return data ? new Transaction(data) : null;
    } catch (error) {
      logger.logError(error, { action: 'findTransactionByTxnId', txnId });
      throw error;
    }
  }

  /**
   * Get transactions by user ID
   */
  static async findByUserId(userId, page = 1, limit = 20) {
    try {
      const offset = (page - 1) * limit;
      
      const { data, error, count } = await supabase
        .from('transactions')
        .select('*', { count: 'exact' })
        .eq('user_id', userId)
        .order('created_at', { ascending: false })
        .range(offset, offset + limit - 1);

      if (error) throw error;

      return {
        transactions: data.map(txn => new Transaction(txn)),
        total: count,
        page,
        limit,
        totalPages: Math.ceil(count / limit)
      };
    } catch (error) {
      logger.logError(error, { action: 'findTransactionsByUserId', userId });
      throw error;
    }
  }

  /**
   * Update transaction status
   */
  async updateStatus(status, metadata = {}) {
    try {
      const { data, error } = await supabase
        .from('transactions')
        .update({
          status,
          metadata: { ...this.metadata, ...metadata },
          updated_at: new Date().toISOString()
        })
        .eq('id', this.id)
        .select()
        .single();

      if (error) throw error;

      this.status = status;
      this.metadata = { ...this.metadata, ...metadata };
      this.updatedAt = data.updated_at;

      logger.logTransaction(this.txnId, 'status_updated', { 
        oldStatus: this.status, 
        newStatus: status 
      });

      return this;
    } catch (error) {
      logger.logError(error, { 
        action: 'updateTransactionStatus', 
        txnId: this.txnId, 
        status 
      });
      throw error;
    }
  }

  /**
   * Update transaction with conversion details
   */
  async updateConversion(convertedAmount, convertedCurrency) {
    try {
      const { data, error } = await supabase
        .from('transactions')
        .update({
          converted_amount: convertedAmount,
          converted_currency: convertedCurrency,
          updated_at: new Date().toISOString()
        })
        .eq('id', this.id)
        .select()
        .single();

      if (error) throw error;

      this.convertedAmount = convertedAmount;
      this.convertedCurrency = convertedCurrency;
      this.updatedAt = data.updated_at;

      logger.logTransaction(this.txnId, 'conversion_updated', {
        originalAmount: this.amount,
        originalCurrency: this.currency,
        convertedAmount,
        convertedCurrency
      });

      return this;
    } catch (error) {
      logger.logError(error, { 
        action: 'updateTransactionConversion', 
        txnId: this.txnId 
      });
      throw error;
    }
  }

  /**
   * Get all transactions (for dashboard)
   */
  static async getAll(filters = {}, page = 1, limit = 50) {
    try {
      const offset = (page - 1) * limit;
      let query = supabase
        .from('transactions')
        .select('*', { count: 'exact' });

      // Apply filters
      if (filters.status) {
        query = query.eq('status', filters.status);
      }
      if (filters.type) {
        query = query.eq('type', filters.type);
      }
      if (filters.startDate) {
        query = query.gte('created_at', filters.startDate);
      }
      if (filters.endDate) {
        query = query.lte('created_at', filters.endDate);
      }
      if (filters.minAmount) {
        query = query.gte('amount', filters.minAmount);
      }
      if (filters.maxAmount) {
        query = query.lte('amount', filters.maxAmount);
      }

      const { data, error, count } = await query
        .order('created_at', { ascending: false })
        .range(offset, offset + limit - 1);

      if (error) throw error;

      return {
        transactions: data.map(txn => new Transaction(txn)),
        total: count,
        page,
        limit,
        totalPages: Math.ceil(count / limit),
        filters
      };
    } catch (error) {
      logger.logError(error, { action: 'getAllTransactions', filters, page, limit });
      throw error;
    }
  }

  /**
   * Get transaction statistics
   */
  static async getStats(period = '24h') {
    try {
      let startDate;
      const now = new Date();
      
      switch (period) {
        case '1h':
          startDate = new Date(now - 60 * 60 * 1000);
          break;
        case '24h':
          startDate = new Date(now - 24 * 60 * 60 * 1000);
          break;
        case '7d':
          startDate = new Date(now - 7 * 24 * 60 * 60 * 1000);
          break;
        case '30d':
          startDate = new Date(now - 30 * 24 * 60 * 60 * 1000);
          break;
        default:
          startDate = new Date(now - 24 * 60 * 60 * 1000);
      }

      const { data, error } = await supabase
        .from('transactions')
        .select('status, type, amount, currency')
        .gte('created_at', startDate.toISOString());

      if (error) throw error;

      const stats = {
        total: data.length,
        totalAmount: 0,
        byStatus: {},
        byType: {},
        byCurrency: {},
        successRate: 0
      };

      data.forEach(txn => {
        stats.totalAmount += parseFloat(txn.amount);
        
        stats.byStatus[txn.status] = (stats.byStatus[txn.status] || 0) + 1;
        stats.byType[txn.type] = (stats.byType[txn.type] || 0) + 1;
        stats.byCurrency[txn.currency] = (stats.byCurrency[txn.currency] || 0) + 1;
      });

      const successfulTxns = stats.byStatus.completed || 0;
      stats.successRate = stats.total > 0 ? (successfulTxns / stats.total * 100).toFixed(2) : 0;

      return stats;
    } catch (error) {
      logger.logError(error, { action: 'getTransactionStats', period });
      throw error;
    }
  }

  /**
   * Get recent transactions for dashboard
   */
  static async getRecent(limit = 10) {
    try {
      const { data, error } = await supabase
        .from('transactions')
        .select('*')
        .order('created_at', { ascending: false })
        .limit(limit);

      if (error) throw error;

      return data.map(txn => new Transaction(txn));
    } catch (error) {
      logger.logError(error, { action: 'getRecentTransactions', limit });
      throw error;
    }
  }

  /**
   * Check if transaction is valid for processing
   */
  canProcess() {
    return this.status === 'pending' && 
           this.amount > 0 && 
           this.userId && 
           this.type;
  }

  /**
   * Get transaction summary for API response
   */
  getSummary() {
    return {
      txnId: this.txnId,
      amount: this.amount,
      currency: this.currency,
      status: this.status,
      type: this.type,
      merchantName: this.merchantName,
      description: this.description,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    };
  }
}

module.exports = Transaction; 

File: backend/src/models/transaction.py
"""
Transaction model for SatuPay Payment Switch
"""

import uuid
from datetime import datetime, timezone
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy import JSON
from src.database.connection import db


class Transaction(db.Model):
    """Transaction model for payment transactions"""

    __tablename__ = 'transactions'

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    txn_id = db.Column(db.String(100), unique=True, nullable=False, index=True)
    user_id = db.Column(UUID(as_uuid=True),
                        db.ForeignKey('users.id'), nullable=True)
    amount = db.Column(db.Numeric(15, 2), nullable=False)
    currency = db.Column(db.String(3), default='MYR')
    original_amount = db.Column(db.Numeric(
        15, 2), nullable=True)  # Before FX conversion
    original_currency = db.Column(db.String(3), nullable=True)
    merchant_id = db.Column(db.String(100), nullable=True)
    merchant_name = db.Column(db.String(255), nullable=True)
    # qr, nfc, online, etc.
    payment_method = db.Column(db.String(50), nullable=False)
    # duitnow, paynow, boost, tng
    payment_rail = db.Column(db.String(50), nullable=False)
    # pending, processing, completed, failed, refunded
    status = db.Column(db.String(20), default='pending')
    qr_code_id = db.Column(UUID(as_uuid=True), db.ForeignKey(
        'qr_codes.id'), nullable=True)
    transaction_metadata = db.Column(JSON, default=dict)
    processed_at = db.Column(db.DateTime, nullable=True)
    completed_at = db.Column(db.DateTime, nullable=True)
    created_at = db.Column(
        db.DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = db.Column(
        db.DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    # Relationships
    plugin_logs = db.relationship(
        'PluginLog', backref='transaction', lazy='dynamic', cascade='all, delete-orphan')

    def __repr__(self):
        return f'<Transaction {self.txn_id}>'

    @staticmethod
    def utc_now():
        """Get current UTC time as timezone-aware datetime"""
        return datetime.now(timezone.utc)

    def to_dict(self):
        """Convert transaction to dictionary"""
        return {
            'id': str(self.id),
            'txn_id': self.txn_id,
            'user_id': str(self.user_id) if self.user_id else None,
            'amount': float(self.amount),
            'currency': self.currency,
            'original_amount': float(self.original_amount) if self.original_amount else None,
            'original_currency': self.original_currency,
            'merchant_id': self.merchant_id,
            'merchant_name': self.merchant_name,
            'payment_method': self.payment_method,
            'payment_rail': self.payment_rail,
            'status': self.status,
            'qr_code_id': str(self.qr_code_id) if self.qr_code_id else None,
            'transaction_metadata': self.transaction_metadata,
            'processed_at': self.processed_at.isoformat() if self.processed_at else None,
            'completed_at': self.completed_at.isoformat() if self.completed_at else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }

    @classmethod
    def create_transaction(cls, txn_id, amount, currency='MYR', **kwargs):
        """Create a new transaction"""
        transaction = cls(
            txn_id=txn_id,
            amount=amount,
            currency=currency,
            **kwargs
        )
        db.session.add(transaction)
        db.session.commit()
        return transaction

    def update_status(self, status, **kwargs):
        """Update transaction status with timestamp"""
        self.status = status
        self.updated_at = self.utc_now()

        if status == 'processing':
            self.processed_at = self.utc_now()
        elif status in ['completed', 'failed', 'refunded']:
            self.completed_at = self.utc_now()

        # Update any additional fields
        for key, value in kwargs.items():
            if hasattr(self, key):
                setattr(self, key, value)

        db.session.commit()

    def add_metadata(self, key, value):
        """Add metadata to transaction"""
        if not self.transaction_metadata:
            self.transaction_metadata = {}

        self.transaction_metadata[key] = value
        db.session.merge(self)  # For JSONB updates
        db.session.commit()

    def get_plugin_logs(self):
        """Get all plugin logs for this transaction"""
        return self.plugin_logs.order_by(PluginLog.created_at).all()

    def get_processing_time(self):
        """Calculate total processing time"""
        if self.completed_at and self.created_at:
            return (self.completed_at - self.created_at).total_seconds()
        return None

    @classmethod
    def get_by_txn_id(cls, txn_id):
        """Get transaction by transaction ID"""
        return cls.query.filter_by(txn_id=txn_id).first()

    @classmethod
    def get_transactions_by_status(cls, status, limit=100):
        """Get transactions by status"""
        return cls.query.filter_by(status=status).limit(limit).all()

    @classmethod
    def get_recent_transactions(cls, limit=50):
        """Get recent transactions"""
        return cls.query.order_by(cls.created_at.desc()).limit(limit).all()

    @classmethod
    def get_transaction_stats(cls):
        """Get transaction statistics"""
        total = cls.query.count()
        pending = cls.query.filter_by(status='pending').count()
        completed = cls.query.filter_by(status='completed').count()
        failed = cls.query.filter_by(status='failed').count()

        # Calculate success rate
        success_rate = (completed / total * 100) if total > 0 else 0

        # Calculate total volume
        volume_result = db.session.query(
            db.func.sum(cls.amount)
        ).filter(cls.status == 'completed').scalar()

        total_volume = float(volume_result) if volume_result else 0.0

        return {
            'total_transactions': total,
            'pending': pending,
            'completed': completed,
            'failed': failed,
            'success_rate': round(success_rate, 2),
            'total_volume': total_volume
        }
